contracts/optimistic-ethereum/OVM/OVM_BondManager.sol

```diff
@@ -0,0 +1,201 @@
+        // The sum of all values inside this map MUST be equal to the
+        // value of `totalClaims`
+        mapping(address => uint256) numClaims;
+    }
+    /// For each pre-state root, there's an array of witnessProviders that must be rewarded
+    /// for posting witnesses
+    mapping(bytes32 => Rewards) public witnessProviders;
+
+    /// Initializes with a ERC20 token to be used for the fidelity bonds
+    /// and with the Address Manager
+    constructor(ERC20 _token, address _libAddressManager)
+        Lib_AddressResolver(_libAddressManager)
+    {
+        owner = msg.sender;
+        token = _token;
+        // ovmFraudVerifier = resolve("OVM_FraudVerifier"); // TODO: Re-enable this
+        ovmCanonicalStateCommitmentChain = resolve("OVM_CanonicalStateCommitmentChain");
+    }
+
+    /// Adds `who` to the list of witnessProviders for the provided `preStateRoot`.
+    function storeWitnessProvider(bytes32 _preStateRoot, address who) public {
+        // The sender must be the transitioner that corresponds to the claimed pre-state root
+        address transitioner = address(iOVM_FraudVerifier(resolve("OVM_FraudVerifier")).getStateTransitioner(_preStateRoot));
+        require(transitioner == msg.sender, Errors.ONLY_TRANSITIONER);
+
+        witnessProviders[_preStateRoot].total += 1;
+        witnessProviders[_preStateRoot].numClaims[who] += 1;
+    }
+
+    /// Slashes + distributes rewards or frees up the sequencer's bond, only called by
+    /// `FraudVerifier.finalizeFraudVerification`
+    function finalize(bytes32 _preStateRoot, uint256 batchIndex, address publisher, uint256 timestamp) public {
+        require(msg.sender == resolve("OVM_FraudVerifier"), Errors.ONLY_FRAUD_VERIFIER);
+        require(witnessProviders[_preStateRoot].canClaim == false, Errors.ALREADY_FINALIZED);
+
+        // allow users to claim from that state root's
+        // pool of collateral (effectively slashing the sequencer)
+        witnessProviders[_preStateRoot].canClaim = true;
+
+        Bond storage bond = bonds[publisher];
+
+        // always slash 1 round of collateral, if possible
+        if (bond.locked >= requiredCollateral) {
+            bond.locked -= requiredCollateral;
+        }
+
+        // if the publisher has a pending withdrawal that's within the published
+        // block's challenge period
+        if (
+            bond.withdrawalTimestamp != 0 && 
+            bond.withdrawalTimestamp <= timestamp + forceInclusionPeriodSeconds &&
+            bond.withdrawing >= requiredCollateral
+        ) {
+            bond.withdrawing -= requiredCollateral;
+        }
+    }
+
+    /// Starts the withdrawal for a publisher
+    function startWithdrawal() public {
+        Bond storage bond = bonds[msg.sender];
+        require(bond.withdrawalTimestamp == 0, Errors.WITHDRAWAL_PENDING);
+        require(bond.locked >= requiredCollateral, Errors.NOT_ENOUGH_COLLATERAL);
+        bond.locked -= requiredCollateral;
```

Comment:
Yeah with above method (checking `bond.withdrawing <= bond.locked`) we can avoid changing `bond.locked` within this function.